Can't use Options or Lists of non-primitive types inside messages when using Protobuf serialization (regression)

I found this out while upgrading the old Metrifier project - at some point after freestyle-rpc 0.3.0, we've lost the ability to handle Lists or Options of non-primitive messages types within an RPC message class. This affects only Protobuf serialization (not Avro).

For example:
```
case class Person(attribs: List[String])
```
works, but
```
case class PersonList(ps: List[Person])
```
doesn't. The exception looks like this:
```
[error] (run-main-0) io.grpc.StatusRuntimeException: CANCELLED: Failed to read message.
[error] io.grpc.StatusRuntimeException: CANCELLED: Failed to read message.
[error] 	at io.grpc.Status.asRuntimeException(Status.java:526)
[error] 	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:467)
[error] 	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:37)
[error] 	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
[error] 	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
[error] 	at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:684)
[error] 	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:37)
[error] 	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
[error] 	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
[error] 	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:391)
[error] 	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:471)
[error] 	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
[error] 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:553)
[error] 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:474)
[error] 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:538)
[error] 	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
[error] 	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:748)
[error] Caused by: java.util.NoSuchElementException: head of empty list
[error] 	at scala.collection.immutable.Nil$.head(List.scala:428)
[error] 	at scala.collection.immutable.Nil$.head(List.scala:425)
[error] 	at pbdirect.PBParserImplicits.$anonfun$requiredParser$1(PBReader.scala:152)
[error] 	at pbdirect.PBParserImplicits.$anonfun$requiredParser$1$adapted(PBReader.scala:141)
[error] 	at pbdirect.LowPriorityPBParserImplicits$$anon$2.parse(PBReader.scala:97)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1(PBReader.scala:106)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1$adapted(PBReader.scala:105)
[error] 	at pbdirect.LowPriorityPBParserImplicits$$anon$2.parse(PBReader.scala:97)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1(PBReader.scala:106)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1$adapted(PBReader.scala:105)
[error] 	at pbdirect.LowPriorityPBParserImplicits$$anon$2.parse(PBReader.scala:97)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1(PBReader.scala:106)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1$adapted(PBReader.scala:105)
[error] 	at pbdirect.LowPriorityPBParserImplicits$$anon$2.parse(PBReader.scala:97)
[error] 	at pbdirect.PBReaderImplicits.$anonfun$prodReader$1(PBReader.scala:41)
[error] 	at pbdirect.LowerPriorityPBReaderImplicits$$anon$1.read(PBReader.scala:17)
[error] 	at pbdirect.PBParserImplicits.$anonfun$requiredParser$1(PBReader.scala:148)
[error] 	at pbdirect.PBParserImplicits.$anonfun$requiredParser$1$adapted(PBReader.scala:141)
[error] 	at pbdirect.LowPriorityPBParserImplicits$$anon$2.parse(PBReader.scala:97)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1(PBReader.scala:106)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1$adapted(PBReader.scala:105)
[error] 	at pbdirect.LowPriorityPBParserImplicits$$anon$2.parse(PBReader.scala:97)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1(PBReader.scala:106)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1$adapted(PBReader.scala:105)
[error] 	at pbdirect.LowPriorityPBParserImplicits$$anon$2.parse(PBReader.scala:97)
[error] 	at pbdirect.PBReaderImplicits.$anonfun$prodReader$1(PBReader.scala:41)
[error] 	at pbdirect.LowerPriorityPBReaderImplicits$$anon$1.read(PBReader.scala:17)
[error] 	at pbdirect.PBParserImplicits.$anonfun$repeatedParser$1(PBReader.scala:134)
[error] 	at pbdirect.PBParserImplicits.$anonfun$repeatedParser$1$adapted(PBReader.scala:127)
[error] 	at pbdirect.LowPriorityPBParserImplicits$$anon$2.parse(PBReader.scala:97)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1(PBReader.scala:106)
[error] 	at pbdirect.LowPriorityPBParserImplicits.$anonfun$consParser$1$adapted(PBReader.scala:105)
[error] 	at pbdirect.LowPriorityPBParserImplicits$$anon$2.parse(PBReader.scala:97)
[error] 	at pbdirect.PBReaderImplicits.$anonfun$prodReader$1(PBReader.scala:41)
[error] 	at pbdirect.LowerPriorityPBReaderImplicits$$anon$1.read(PBReader.scala:17)
[error] 	at pbdirect.PBParserImplicits.$anonfun$requiredParser$1(PBReader.scala:148)
[error] 	at pbdirect.PBParserImplicits.$anonfun$requiredParser$1$adapted(PBReader.scala:141)
[error] 	at pbdirect.LowPriorityPBParserImplicits$$anon$2.parse(PBReader.scala:97)
[error] 	at pbdirect.package$PBParserOps$.pbTo$extension(package.scala:26)
[error] 	at freestyle.rpc.internal.encoders$pbd$$anon$1.parse(encoders.scala:71)
[error] 	at io.grpc.MethodDescriptor.parseResponse(MethodDescriptor.java:266)
[error] 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:526)
[error] 	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
[error] 	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:748)
```

--- a/modules/internal/src/main/scala/service.scala
+++ b/modules/internal/src/main/scala/service.scala
@@ -157,11 +157,15 @@ object serviceImpl {
 
       val encodersImport = serializationType match {
         case Protobuf =>
-          q"import _root_.freestyle.rpc.internal.encoders.pbd._"
+          List(
+            q"import _root_.cats.instances.list._",
+            q"import _root_.cats.instances.option._",
+            q"import _root_.freestyle.rpc.internal.encoders.pbd._"
+          )
         case Avro =>
-          q"import _root_.freestyle.rpc.internal.encoders.avro._"
+          List(q"import _root_.freestyle.rpc.internal.encoders.avro._")
         case AvroWithSchema =>
-          q"import _root_.freestyle.rpc.internal.encoders.avrowithschema._"
+          List(q"import _root_.freestyle.rpc.internal.encoders.avrowithschema._")
       }
 
       val methodDescriptors: List[Tree] = rpcRequests.map(_.methodDescriptor)
@@ -390,8 +394,7 @@ object serviceImpl {
           Template(
             companion.impl.parents,
             companion.impl.self,
-            companion.impl.body ++ service.imports ++ service.methodDescriptors ++ List(
-              service.encodersImport,
+            companion.impl.body ++ service.imports ++ service.methodDescriptors ++ service.encodersImport ++ List(
               service.bindService,
               service.clientClass,
               service.client,

--- a/modules/server/src/test/scala/protocol/RPCProtoProducts.scala
+++ b/modules/server/src/test/scala/protocol/RPCProtoProducts.scala
@@ -1,0 +1,199 @@
+/*
+ * Copyright 2017-2018 47 Degrees, LLC. <http://www.47deg.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package freestyle.rpc
+package protocol
+
+import cats.Applicative
+import cats.syntax.applicative._
+import freestyle.rpc.common._
+import freestyle.rpc.testing.servers.withServerChannel
+import org.scalatest._
+import org.scalacheck.Prop._
+import org.scalatest.prop.Checkers
+
+class RPCProtoProducts extends RpcBaseTestSuite with BeforeAndAfterAll with Checkers {
+
+  object RPCService {
+
+    case class MyParam(value: String)
+
+    case class RequestOption(param1: Option[MyParam])
+
+    case class ResponseOption(param1: Option[String], param2: Boolean)
+
+    case class RequestList(param1: List[MyParam])
+
+    case class ResponseList(param1: List[String], param2: Boolean)
+
+    @service(Protobuf)
+    trait ProtoRPCServiceDef[F[_]] {
+      def optionProto(req: RequestOption): F[ResponseOption]
+      def listProto(req: RequestList): F[ResponseList]
+    }
+
+    @service(Avro)
+    trait AvroRPCServiceDef[F[_]] {
+      def optionAvro(req: RequestOption): F[ResponseOption]
+      def listAvro(req: RequestList): F[ResponseList]
+    }
+
+    @service(AvroWithSchema)
+    trait AvroWithSchemaRPCServiceDef[F[_]] {
+      def optionAvroWithSchema(req: RequestOption): F[ResponseOption]
+      def listAvroWithSchema(req: RequestList): F[ResponseList]
+    }
+
+    class RPCServiceDefImpl[F[_]: Applicative]
+        extends ProtoRPCServiceDef[F]
+        with AvroRPCServiceDef[F]
+        with AvroWithSchemaRPCServiceDef[F] {
+
+      def optionProto(req: RequestOption): F[ResponseOption] =
+        ResponseOption(req.param1.map(_.value), true).pure
+      def optionAvro(req: RequestOption): F[ResponseOption] =
+        ResponseOption(req.param1.map(_.value), true).pure
+      def optionAvroWithSchema(req: RequestOption): F[ResponseOption] =
+        ResponseOption(req.param1.map(_.value), true).pure
+
+      def listProto(req: RequestList): F[ResponseList] =
+        ResponseList(req.param1.map(_.value), true).pure
+      def listAvro(req: RequestList): F[ResponseList] =
+        ResponseList(req.param1.map(_.value), true).pure
+      def listAvroWithSchema(req: RequestList): F[ResponseList] =
+        ResponseList(req.param1.map(_.value), true).pure
+    }
+
+  }
+
+  "A RPC server" should {
+
+    import RPCService._
+    import monix.execution.Scheduler.Implicits.global
+
+    implicit val H: RPCServiceDefImpl[ConcurrentMonad] =
+      new RPCServiceDefImpl[ConcurrentMonad]
+
+    "be able to serialize and deserialize Options in the request/response using proto format" in {
+
+      withServerChannel(ProtoRPCServiceDef.bindService[ConcurrentMonad]) { sc =>
+        val client: ProtoRPCServiceDef.Client[ConcurrentMonad] =
+          ProtoRPCServiceDef.clientFromChannel[ConcurrentMonad](sc.channel)
+
+        check {
+          forAll { maybeString: Option[String] =>
+            client
+              .optionProto(RequestOption(maybeString.map(MyParam)))
+              .unsafeRunSync() == ResponseOption(maybeString, true)
+          }
+        }
+
+      }
+
+    }
+
+    "be able to serialize and deserialize Options in the request/response using avro format" in {
+
+      withServerChannel(AvroRPCServiceDef.bindService[ConcurrentMonad]) { sc =>
+        val client: AvroRPCServiceDef.Client[ConcurrentMonad] =
+          AvroRPCServiceDef.clientFromChannel[ConcurrentMonad](sc.channel)
+
+        check {
+          forAll { maybeString: Option[String] =>
+            client
+              .optionAvro(RequestOption(maybeString.map(MyParam)))
+              .unsafeRunSync() == ResponseOption(maybeString, true)
+          }
+        }
+
+      }
+
+    }
+
+    "be able to serialize and deserialize Options in the request/response using avro with schema format" in {
+
+      withServerChannel(AvroWithSchemaRPCServiceDef.bindService[ConcurrentMonad]) { sc =>
+        val client: AvroWithSchemaRPCServiceDef.Client[ConcurrentMonad] =
+          AvroWithSchemaRPCServiceDef.clientFromChannel[ConcurrentMonad](sc.channel)
+
+        check {
+          forAll { maybeString: Option[String] =>
+            client
+              .optionAvroWithSchema(RequestOption(maybeString.map(MyParam)))
+              .unsafeRunSync() == ResponseOption(maybeString, true)
+          }
+        }
+
+      }
+
+    }
+
+    "be able to serialize and deserialize Lists in the request/response using proto format" in {
+
+      withServerChannel(ProtoRPCServiceDef.bindService[ConcurrentMonad]) { sc =>
+        val client: ProtoRPCServiceDef.Client[ConcurrentMonad] =
+          ProtoRPCServiceDef.clientFromChannel[ConcurrentMonad](sc.channel)
+
+        check {
+          forAll { list: List[String] =>
+            client
+              .listProto(RequestList(list.map(MyParam)))
+              .unsafeRunSync() == ResponseList(list, true)
+          }
+        }
+
+      }
+
+    }
+
+    "be able to serialize and deserialize Lists in the request/response using avro format" in {
+
+      withServerChannel(AvroRPCServiceDef.bindService[ConcurrentMonad]) { sc =>
+        val client: AvroRPCServiceDef.Client[ConcurrentMonad] =
+          AvroRPCServiceDef.clientFromChannel[ConcurrentMonad](sc.channel)
+
+        check {
+          forAll { list: List[String] =>
+            client
+              .listAvro(RequestList(list.map(MyParam)))
+              .unsafeRunSync() == ResponseList(list, true)
+          }
+        }
+
+      }
+
+    }
+
+    "be able to serialize and deserialize Lists in the request/response using avro with schema format" in {
+
+      withServerChannel(AvroWithSchemaRPCServiceDef.bindService[ConcurrentMonad]) { sc =>
+        val client: AvroWithSchemaRPCServiceDef.Client[ConcurrentMonad] =
+          AvroWithSchemaRPCServiceDef.clientFromChannel[ConcurrentMonad](sc.channel)
+
+        check {
+          forAll { list: List[String] =>
+            client
+              .listAvroWithSchema(RequestList(list.map(MyParam)))
+              .unsafeRunSync() == ResponseList(list, true)
+          }
+        }
+
+      }
+
+    }
+  }
+
+}

--- a/version.sbt
+++ b/version.sbt
@@ -1,1 +1,1 @@
-version in ThisBuild := "0.14.1-SNAPSHOT"
+version in ThisBuild := "0.14.1"

